### PR TITLE
fast path for declared_generic_bounds_from_env

### PR DESCRIPTION
**Pull request by aliemjay**
~2% perf gain for diesel

### CODE REVIEW

**Review by compiler-errors — APPROVED** [review]
r=me after perf

File: compiler/rustc_infer/src/infer/outlives/verify.rs [hunk 1]
```diff
@@ -245,10 +245,19 @@ impl<'cx, 'tcx> VerifyBoundCx<'cx, 'tcx> {
                     "declared_generic_bounds_from_env_for_erased_ty: region_bound_pair = {:?}",
                     (r, p)
                 );
+                // Fast path for the common case.
+                match (erased_ty.kind(), &p) {
+                    (ty::Alias(_, a1), GenericKind::Alias(a2)) if a1.def_id != a2.def_id => {
```
Comment (compiler-errors) [inline compiler/rustc_infer/src/infer/outlives/verify.rs:250]: This *could* use a comment saying that we expect all preds here to have only rigid projections.

File: compiler/rustc_infer/src/infer/outlives/verify.rs [hunk 2]
```diff
@@ -245,10 +245,19 @@ impl<'cx, 'tcx> VerifyBoundCx<'cx, 'tcx> {
...
+                        return None;
+                    }
+                    (ty::Param(p1), GenericKind::Param(p2)) if p1 != p2 => return None,
+                    _ => {}
```
Comment (aliemjay) [inline compiler/rustc_infer/src/infer/outlives/verify.rs:254]:
```suggestion
                    (ty::Alias(_, a1), GenericKind::Alias(a2)) if a1.def_id == a2.def_id =>{}
                    (ty::Param(p1), GenericKind::Param(p2)) if p1 == p2 => {}
                    _ => return None,
```
This should cover more cases.

**Comment by aliemjay** [comment]
Added a comment about rigid aliase types and implemented an [alternative approach](https://github.com/rust-lang/rust/pull/119084#discussion_r1430389025)

@rusbot review

File: compiler/rustc_infer/src/infer/outlives/verify.rs (hunk 2 above)
Comment (aliemjay) [reply compiler/rustc_infer/src/infer/outlives/verify.rs]: This is missing the case of `GenericKind::Placeholder`.

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
r=me

**Comment by aliemjay** [comment]
@bor r=compiler-errors  rollup=never

**Comment by nnethercote** [comment]
@aliemjay: how did you find this improvement? Profiling, looking through the code, something else? I'm curious :)

**Comment by aliemjay** [comment]
> @aliemjay: how did you find this improvement? Profiling, looking through the code, something else? I'm curious :)

by callgrind profiles, looking at code that I have some expertise in so that I can easily catch optimazation opportunities.

**Pull request merged**